runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/Scheduler.java

```diff
@@ -0,0 +1,222 @@
+
+  private static final Histogram scanTime = Histogram.build().buckets(1, 5, 10, 30, 60, 300, 600, 3600)
+      .name("miso_runscanner_directory_scan_time").help("Time to scan the run directories in seconds.").register();
+
+  // The paths that threw an exception while processing.
+  private final Set<File> failed = new ConcurrentSkipListSet<>();
+  // The paths for which we have a notification to send.
+  private final Map<File, NotificationDto> finishedWork = new ConcurrentHashMap<>();
+
+  // The paths that are current being processed (and the corresponding processor).
+  private final Set<File> processing = new ConcurrentSkipListSet<>();
+
+  // The directories that contain run directories that need to be scanned and the processors for those runs.
+  private Map<File, Configuration> roots = new ConcurrentHashMap<>();
+
+  private ScheduledFuture<?> scanDirectoriesFuture = null;
+
+  private boolean scanningNow = false;
+
+  private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+  private final ExecutorService workPool = Executors.newWorkStealingPool();
+
+  // The paths that need to be processed (and the corresponding processor).
+  private final Set<File> workToDo = new ConcurrentSkipListSet<>();
+
+  public Stream<Map.Entry<File, NotificationDto>> finished() {
+    return finishedWork.entrySet().stream();
+  }
+
+  public Iterable<Configuration> getConfiguration() {
+    return roots.values();
+  }
+
+  public Set<File> getCurrentWork() {
+    return processing;
+  }
+
+  public Set<File> getFailedDirectories() {
+    return failed;
+  }
+
+  public Set<File> getFinishedDirectories() {
+    return finishedWork.keySet();
+  }
+
+  public Set<File> getRoots() {
+    return roots.keySet();
+  }
+
+  public Set<File> getScheduledWork() {
+    return workToDo;
+  }
+
+  public boolean isScanningEnabled() {
+    return scanDirectoriesFuture != null;
+  }
+
+  public boolean isScanningNow() {
+    return scanningNow;
+  }
+
+  @Value("${runscanner.configFile}")
+  public void setConfigurationFile(String filename) {
+    ObjectMapper mapper = new ObjectMapper();
+    try {
+      roots = Arrays.stream(mapper.readValue(new File(filename), SuppliedDirectoryConfig[].class)).map(source -> {
+        Configuration destination = new Configuration();
+        destination.setPath(new File(source.getPath()));
+        destination.setTimeZone(TimeZone.getTimeZone(source.getTimeZone()));
+        for (RunProcessor processor : RunProcessor.INSTANCES) {
+          if (processor.getName().equals(source.getName()) && processor.getPlatformType() == source.getPlatformType()) {
+            destination.setProcessor(processor);
+            break;
+          }
+        }
+        return destination;
+      }).collect(Collectors.toConcurrentMap(Configuration::getPath, Function.identity()));
+    } catch (IOException e) {
+      log.error("Confguration is bad.", e);
```

Comment:
I've changed this logic. I think the servlet should keep running even if the configuration is bad. It should report bad configuration and scan for a changed configuration. If I correct the configuration on disk, having the servlet fail on start up only adds steps for me having to fix it.

Also, consider the configuration where the configuration lives in git. The servlet just picks up a new configuration as it gets pulled from git. If junk is checked into git, the servlet either continues to use the old config, or does nothing until a new one comes a long. It can report the failure to the monitoring system.